[test, init] DNS seed querying logic

amitiuttarwar opened the pull request:
This PR adds a DNS seed to the regtest chain params to enable testing the DNS seed querying logic of `CConnman::ThreadDNSAddressSeed` and relevant startup parameters. Adds coverage for the changes in #22013 (and then some).

The main behavioral change to bitcoind is that this PR disallows starting up with conflicting parameters for `-dnsseed` and `-forcednsseed`. 

The tests include:
* parameter interactions of different combinations of `-connect`, `-dnsseed` and `-forcednsseed`
* the delay before querying DNS seeds depending on how many addresses are in the addrman 
* the behavior of `-forcednsseed` 
* skipping DNS querying if we have outbound full relay connections & not block-relay-only connections 

Huge props to mzumsande for identifying the timing technique for testing successful connections before running `ThreadDNSAddressSeed` 🙌🏽

amitiuttarwar committed:
[style] Small style improvements to DNS parameters

amitiuttarwar committed:
[test] Test the interactions between -connect and -dnsseed

amitiuttarwar committed:
[test] Test -forcednsseed causes querying DNS seeds

amitiuttarwar committed:
[test] Test the delay before querying DNS seeds

When starting up with a populated addrman, ThreadDNSAddressSeed adds a delay
during which time the node may be able to connect to some peers. This commit
tests the delay changes based on the number of addresses in the addrman.

amitiuttarwar committed:
[init] Disallow starting up with conflicting paramters for -dnsseed and -forcednsseed

-dnsseed determines whether we run ThreadDNSAddressSeed and potentially query
the DNS seeds for addresses. -forcednsseed tells the node to force querying the
DNS seeds even if we have sufficient addresses or current connections.

This commit disallows starting up with explicitly conflicting parameters.

amitiuttarwar committed:
[test] Test the interactions between -forcednsseed and -dnsseed

Test that passing conflicting parameters for the two causes a startup error.
This logic also impacts -connect, which soft sets -dnsseed, so add a test for
that too.

amitiuttarwar committed:
[test] Introduce test logic to query DNS seeds

This commit introduces a DNS seed to the regest chain params in order to add
coverage to the DNS querying logic.

The first test checks that we do not query DNS seeds if we are able to
succesfully connect to 2 outbound connections. Since we participate in ADDR
relay with those connections, including sending a GETADDR message during the
VERSION handshake, querying the DNS seeds is unnecessary.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
[test] Test logic to query DNS seeds with block-relay-only connections

When a node is able to properly shutdown, it will persist its block-relay-only
connections to the addrman. On startup, it will attempt to reconnect to these
anchors. Since block-relay-only connections do not participate in ADDR relay,
succesful connections are insufficient to skip querying the DNS seeds.

This test fails prior to the changes in #22013.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

lsilva01 reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/22098/commits/75ee018f9b8371cfb16bd3ad8f9491d386d9c238 on Ubuntu 20.04.

willcl-ark reviewed (approved):
This looks like a good change which clears up the behaviour of some contradictory flag combinations and more thoroughly tests the dns fallback behaviour. tACK each commit in the series.

Logic added in ea0ad085e80106aee305628f53d4b7409ed9e44c is exercised by the new test `test/functional/p2p_dns_seeds.py`

laanwj commented:
Thanks for adding tests for this functionality. This goes as far as can be reasonably expected without emulating a DNS server in the tests :sweat_smile: 

Code review ACK 75ee018f9b8371cfb16bd3ad8f9491d386d9c238

mzumsande commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -17,6 +17,7 @@ def set_test_params(self):
     def run_test(self):
         self.existing_outbound_connections_test()
         self.existing_block_relay_connections_test()
+        self.init_arg_tests()
```
The test runs quite long (much waiting). I don't think it's possible to prevent this (mocking time doesn't seem possible) but running `init_arg_tests` before `existing_outbound_connections_test` would save one 11 second wait because addrman would still be empty.

amitiuttarwar replied:
good call! changed the ordering

mzumsande reviewed (commented):
Concept ACK. Some non-critical suggestions below, although maybe it's worthwile speeding the test up a bit.

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,124 @@
+    def existing_block_relay_connections_test(self):
+        # Make sure addrman is populated to enter the conditional where we
+        # delay and potentially skip DNS seeding.
+        self.nodes[0].addpeeraddress("192.0.0.8", 8333)
```
ah, totally. this line is unnecessary when running the test together, but I added it so the test setup would not be dependent on `existing_outbound_connections_test`. would it better if I change the IP address to be something new?

mzumsande replied:
oh, I see. No, I don't think that is necessary.

amitiuttarwar replied:
didn't change it, but added a comment to clarify.

jnewbery commented on src/net.h:
```diff
@@ -79,9 +79,9 @@ static const int64_t DEFAULT_PEER_CONNECT_TIMEOUT = 60;
-static const bool DEFAULT_FORCEDNSSEED = false;
-static const bool DEFAULT_DNSSEED = true;
-static const bool DEFAULT_FIXEDSEEDS = true;
+static constexpr bool DEFAULT_FORCEDNSSEED = false;
+static constexpr bool DEFAULT_DNSSEED = true;
+static constexpr bool DEFAULT_FIXEDSEEDS = true;
```
If you're touching these anyway...

```suggestion
static constexpr bool DEFAULT_FORCEDNSSEED{false};
static constexpr bool DEFAULT_DNSSEED{true};
static constexpr bool DEFAULT_FIXEDSEEDS{true};
```

amitiuttarwar replied:
done

jnewbery reviewed (commented):
ACK 75ee018f9b8371cfb16bd3ad8f9491d386d9c238

Thanks for adding these tests!

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,124 @@
+            self.restart_node(0, ["-forcednsseed", "-dnsseed=1"])
+
+        # Restore default for subsequent tests
+        self.restart_node(0, ["-forcednsseed=0"])
```
ah I see! I thought the args forwarded between restarts, but I see now its just the initial setup args that get retained. updated these calls.

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,124 @@
+        )
+
+        # Restore default bitcoind settings
+        self.restart_node(0, ["-forcednsseed=0", "-dnsseed=1", "-connect=0"])
```
done

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,124 @@
+            self.restart_node(0)
+
+        # Populate addrman with > 1000 addresses
+        for i in range(1500):
```
nice! in addition to avoiding the magic number, it means the test has the flexibility to end earlier if we meet the required number of addresses in addrman.

amitiuttarwar commented:
thank you for the reviews @lsilva01, @willcl-ark, @laanwj, @mzumsande & @jnewbery! 

I've incorporated all the review comments ([diff](https://github.com/bitcoin/bitcoin/compare/75ee018f9b8371cfb16bd3ad8f9491d386d9c238..6108630a9e6a250922ab8189172476927223b4e8)).

I also changed the regtest dns seed to be `dummySeed.invalid` ([diff](https://github.com/bitcoin/bitcoin/compare/6108630a9e6a250922ab8189172476927223b4e8..a8834fa0513d6bfcd4dfc759bbb841c02fcf8517)). Since `.invalid` is reserved to ensure no real results will be returned ([wiki](https://en.wikipedia.org/wiki/.invalid), [IETF RFC](https://datatracker.ietf.org/doc/html/rfc2606)), it seems like a good fit here.

jnewbery commented:
Code review ACK a8834fa

kristapsk reviewed (approved):
ACK a8834fa0513d6bfcd4dfc759bbb841c02fcf8517

mzumsande commented:
Code Review ACK a8834fa0513d6bfcd4dfc759bbb841c02fcf8517

practicalswift commented on src/chainparams.cpp:
```diff
@@ -435,7 +435,8 @@ class CRegTestParams : public CChainParams {
         vFixedSeeds.clear(); //!< Regtest mode doesn't have any fixed seeds.
-        vSeeds.clear();      //!< Regtest mode doesn't have any DNS seeds.
+        vSeeds.clear();
+        vSeeds.emplace_back("dummySeed.invalid");
```
Could use a FQDN such as say `dummy-seed.invalid.` (note the ending dot) be used here to make it unambiguous and thus make sure we don't risk causing a DNS lookup for `dummySeed.invalid.<local domain name as specified in /etc/resolv.conf>`)?

amitiuttarwar replied:
good call, updated to `dummySeed.invalid.`

practicalswift reviewed (changes requested):
Concept ACK

Thanks for improving testing!

I left some review comments on small changes that could further reduce the risk of accidental external communication :)

mzumsande commented:
This probably needs a rebase / exception added for the new test after #22490 was merged.

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,127 @@
+    def existing_outbound_connections_test(self):
+        # Make sure addrman is populated to enter the conditional where we
+        # delay and potentially skip DNS seeding.
+        self.nodes[0].addpeeraddress("192.0.0.8", 8333)
```
hmmm, I tried this out but unfortunately since the `CNetAddr` code specifically recognizes these as invalid addresses ([link](https://github.com/bitcoin/bitcoin/blob/da1c0c64fd094880712d1c4167ad9eb3bb6ffcc6/src/netaddress.cpp#L345-L347)), it prevents being able to test the functionality.

practicalswift replied:
Got it! Then `192.0.0.8` (the IPv4 dummy address per [RFC 7600](https://www.rfc-editor.org/rfc/rfc7600.html)) is probably a good alternative to `192.0.2.0/24`.

I checked the [IANA IPv4 Special-Purpose Address Registry](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml) and `192.0.0.8` (like `192.0.2.0/24`) is guaranteed to not be in use as valid destination address,.

amitiuttarwar replied:
awesome, thanks!

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,127 @@
+        self.log.info("Check that setting -connect disables -dnsseed by default")
+        self.nodes[0].stop_node()
+        with(self.nodes[0].assert_debug_log(expected_msgs=["DNS seeding disabled"])):
+            self.start_node(0, ["-connect=fakenodeaddr"])
```
ah cool, thanks for the suggestion! I updated all 3 uses of `-connect=` to use this FQDN.

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,127 @@
+        self.nodes[0].stop_node()
+        self.nodes[0].assert_start_raises_init_error(
+            expected_msg="Error: Cannot set -forcednsseed to true when setting -dnsseed to false.",
+            extra_args=["-forcednsseed=1", "-connect=fakenodeaddr"],
```
done

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,127 @@
+        # delay and potentially skip DNS seeding. No-op when run after
+        # existing_outbound_connections_test.
+        self.nodes[0].addpeeraddress("192.0.0.8", 8333)
+
```
responded https://github.com/bitcoin/bitcoin/pull/22098#discussion_r680131932. let's continue the convo there

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,127 @@
+    def init_arg_tests(self):
+        self.log.info("Check that setting -connect disables -dnsseed by default")
+        self.nodes[0].stop_node()
+        with(self.nodes[0].assert_debug_log(expected_msgs=["DNS seeding disabled"])):
```
updated all `with(...):` to be `with ...:`

amitiuttarwar commented:
thanks for the reviews @kristapsk & @practicalswift! & for the re-reviews @jnewbery & @mzumsande ☺️

@practicalswift 
I took all your suggestions to further reduce changes of external communication except for being able to apply the IP addresses reserved for testing. Thanks again for these! I learned about how FQDNs work. 

@mzumsande 
as I mentioned over here https://github.com/bitcoin/bitcoin/pull/22490#issuecomment-890039860, I surprisingly don't need to rebase or add an exception! 
>  I realized that test doesn't actually need rebase or an exception added, because passing in something like self.start_node(0, ["-connect=fakenodeaddr"]) from the individual test appends the -connect arg as a command-line argument, which overrules what's been set in the .conf file.

I locally rebased & confirmed that the tests still pass, but opted not to push the rebase here to make it easier to re-review. But lmk if you think it would be better to rebase & set `self.disable_autoconnect` and I'm happy to do so.

practicalswift commented:
> I took all your suggestions to further reduce changes of external communication except for being able to apply the IP addresses reserved for testing. Thanks again for these! I learned about how FQDNs work.

Excellent!

It is outside of the scope of this PR but I think it would make sense to err on the safe side and specify all seed domain names with an ending dot to make them unambiguous.

In other words …

```c++
    vSeeds.emplace_back("seed.bitcoin.sipa.be."); // Pieter Wuille, only supports x1, x5, x9, and xd
```

… instead of the current …

```c++
    vSeeds.emplace_back("seed.bitcoin.sipa.be"); // Pieter Wuille, only supports x1, x5, x9, and xd
```

The former is guaranteed to be interpreted as `seed.bitcoin.sipa.be.` whereas the latter may be interpreted as say`seed.bitcoin.sipa.be.megacorp.com.` depending on the content of the user's `/etc/resolv.conf`.

The case I'm thinking about is if `search megacorp.com` is specified in `/etc/resolv.conf` and say `options ndots:5` is used.

Looking up `seed.bitcoin.sipa.be` would then result in the following DNS traffic:

```
IP <src> > <resolver>.53: A? seed.bitcoin.sipa.be.megacorp.com.
IP <resolver>.53 > <src>: NXDomain
IP <src> > <resolver>.53: A? seed.bitcoin.sipa.be.
IP <resolver>.53 > <src>: Some valid response.
```

In other words `seed.bitcoin.sipa.be.megacorp.com.` would be tried before `seed.bitcoin.sipa.be.` :(

I haven't tried it myself but if I'm thinking correctly here then adding `search localtest.me` and `options ndots:15` to `/etc/resolv.conf` should break DNS seeding since the DNS server for `localtest.me` will return `127.0.0.1` for `*.localtest.me`.

I guess an alternative way to fix this could be to add the ending dot (if missing) when building the host string:

```patch
diff --git a/src/net.cpp b/src/net.cpp
index 073643fb7..c68c497ea 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1714,7 +1714,7 @@ void CConnman::ThreadDNSAddressSeed()
             std::vector<CNetAddr> vIPs;
             std::vector<CAddress> vAdd;
             ServiceFlags requiredServiceBits = GetDesirableServiceFlags(NODE_NONE);
-            std::string host = strprintf("x%x.%s", requiredServiceBits, seed);
+            std::string host = strprintf("x%x.%s.", requiredServiceBits, seed);
             CNetAddr resolveSource;
             if (!resolveSource.SetInternal(host)) {
                 continue;
```

---

Some background context from https://bugzilla.mozilla.org/show_bug.cgi?id=134402:

```
The BSD Unix implementation of the name resolver established numerous
conventions that have become defacto standards for Unix and Linux, but are
not formally Internet standards, and are not commonly found on other 
platforms, such as Windows.  These include such things as:

…

b) a trailing dot tells BSD's DNS name resolver not to try appending domain 
name suffixes.  E.g. for a client in the mozilla.com domain, the DNS name 
foo.bar can be interpreted as foo.bar.mozilla.com as well as foo.bar, but 
foo.bar. is only interpreted as foo.bar and not as foo.bar.mozilla.com.  
These are only conventions of some OSes, and not Internet standards.  
```

amitiuttarwar commented:
all review comments have been addressed. 

> specify all seed domain names with an ending dot to make them unambiguous.. 

@practicalswift, very interesting! thanks for digging in & sharing your findings! I'll have to learn more myself but this seems like a good idea. Shall we open an issue?

mzumsande commented:
ACK 82b6f89819e55af26f5264678e0f93052934bcb3

>I locally rebased & confirmed that the tests still pass, but opted not to push the rebase here to make it easier to re-review.  But lmk if you think it would be better to rebase & set self.disable_autoconnect and I'm happy to do so.

not necessary, I didn't realize that it would work, but your explanation makes sense.

practicalswift commented:
> > The former is guaranteed to be interpreted as `seed.bitcoin.sipa.be.` whereas the latter may be interpreted as say`seed.bitcoin.sipa.be.megacorp.com.` depending on the content of the user's `/etc/resolv.conf`.
>
> @practicalswift, very interesting! thanks for digging in & sharing your findings! I'll have to learn more myself but this seems like a good idea. Shall we open an issue?

Feel free to open an issue if you feel for it: I'm afraid I won't be able to find time to work on this any time soon, but I'd be glad to help out with review :)

An `/etc/resolv.conf` file with this content should be sufficient to reproduce the issue (if my theory holds: I haven't had time to check beyond reading the source!):

```
search localtest.me
options ndots:15
nameserver 8.8.8.8
nameserver 8.8.4.4
```

`8.8.8.8` and `8.8.4.4` are Google Public DNS, and `localtest.me` is a domain name setup with a DNS server that will respond with `127.0.0.1` for all queries including say `seed.bitcoin.sipa.be.localtest.me`.

practicalswift commented:
> > > The former is guaranteed to be interpreted as `seed.bitcoin.sipa.be.` whereas the latter may be interpreted as say`seed.bitcoin.sipa.be.megacorp.com.` depending on the content of the user's `/etc/resolv.conf`.
> > 
> > 
> > @practicalswift, very interesting! thanks for digging in & sharing your findings! I'll have to learn more myself but this seems like a good idea. Shall we open an issue?
> 
> Feel free to open an issue if you feel for it: I'm afraid I won't be able to find time to work on this any time soon, but I'd be glad to help out with review :)
> 
> An `/etc/resolv.conf` file with this content should be sufficient to reproduce the issue (if my theory holds: I haven't had time to check beyond reading the source!):

A simpler way that should reproduce the issue without having to make any changes to `/etc/resolv.conf`:

```
$ RES_OPTIONS="options ndots:15" LOCALDOMAIN="localtest.me" src/bitcoind
```

If my thinking is correct that should make the DNS seeding queries return `127.0.0.1` due to `seed.bitcoin.sipa.be` being interpreted as `seed.bitcoin.sipa.be.localtest.me`, etc.

jnewbery commented:
reACK 82b6f89819e55af26f5264678e0f93052934bcb3

amitiuttarwar commented on test/functional/p2p_dns_seeds.py:
```diff
@@ -0,0 +1,124 @@
+        self.log.info("Check that setting -connect disables -dnsseed by default")
+        self.nodes[0].stop_node()
+        with(self.nodes[0].assert_debug_log(expected_msgs=["DNS seeding disabled"])):
+            self.start_node(0, ["-connect"])
```
oops, forgot to comment / resolve, but this has been fixed for all the `-connect` calls

fanquake commented:
The DNS querying logic and related config options certainly are somewhat convoluted / confusing. This is not helped by the fact that certain options (soft)-set other options, or sometimes behave un-intuitively. I think there is definitely scope to improve this code further going forward.

fanquake commented:
cc @ajtowns
